Adding cycler

GitToby opened the pull request:
## Changes
 - Fixed #43 & added tests 🐛 
 - Added Cycler Params class for genetically optimizing an input sequence 
 - Added Docs for Cycler Params & example
 - Added a Seeding wrapper util method for playing the "same" stochastic opponent multiple times

GitToby committed:
Added initial Cycler & cycler tests

GitToby committed:
changing parameters and adding integration test

GitToby committed:
fixed the windows multi threading issue in an ugly way

to fix details see todo comment

GitToby committed:
added integration test functionality

GitToby committed:
Added docstrings

GitToby committed:
Added docstrings, and extra content for custom initial populations

GitToby committed:
edited the bug for increasing pop sizes

GitToby committed:
Added integration test for pop sizes & seeding players utility

drvinceknight commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
+numpy
```
No need for this, `numpy` is a req of axelrod.

drvinceknight commented on src/axelrod_dojo/utils.py:
```diff
@@ -194,3 +197,12 @@ def load_params(params_class, filename, num):
     return best_params
 
+
+def get_seeded_opponent_instance(player_class):
```
No need for this in here @GitToby. (This is to be used outside of the dojo)

drvinceknight commented on tests/integration/test_cycler_integration.py:
```diff
@@ -0,0 +1,70 @@
+        temp_file.close()
+
+
+class test_seeding_function(unittest.TestCase):
```
Can remove this as well.

drvinceknight commented on tests/integration/test_population_sizes.py:
```diff
@@ -0,0 +1,69 @@
+
+        population = axl_dojo.Population(params_class=axl_dojo.CyclerParams,
+                                         params_kwargs=cycler_kwargs,
+                                         size=POPULATION_SIZE,
```
No need for a module wide variable, can just keep this as `population_size` on line 52.

drvinceknight reviewed (changes requested):
Some initial comments @GitToby, can you also add a short paragraph or two about the Cycler in here: https://github.com/Axelrod-Python/axelrod-dojo/blob/master/docs/background/genetic_algorithm.rst

drvinceknight commented:
(As a headsup, @marcharper, Toby is a Cardiff undergrad doing a project with @Nikoleta-v3 and I.)

drvinceknight commented:
> can you also add a short paragraph or two about the Cycler in here: https://github.com/Axelrod-Python/axelrod-dojo/blob/master/docs/background/genetic_algorithm.rst

Ignore that: somehow missed that you're already done it. Nice job @GitToby.

There's a merge conflict with master, we can fix that on Monday. :+1:

GitToby committed:
removed seeding player & windows checks.

GitToby committed:
removed numpy from the requirements

GitToby committed:
removed global population size var

GitToby committed:
adding multi threading fix

GitToby committed:
Merge branch 'master' into Adding_Cycler

GitToby committed:
Outputter class changed + tests

GitToby committed:
Merge branch 'Adding_Cycler' of https://github.com/GitToby/axelrod-dojo into Adding_Cycler

GitToby committed:
Crossover & tests changed; added header to CSV printing

GitToby committed:
Crossover & tests changed; add header to CSV printing; added travis -v

GitToby commented:
The build is failing from tests in other areas than the Cycler: 
```
$ coverage report -m
Name                                                         Stmts   Miss  Cover   Missing
------------------------------------------------------------------------------------------
src/axelrod_dojo/__init__.py                                     8      0   100%
src/axelrod_dojo/algorithms/__init__.py                          0      0   100%
src/axelrod_dojo/algorithms/genetic_algorithm.py                89      2    98%   38, 128
src/axelrod_dojo/algorithms/particle_swarm_optimization.py      35      2    94%   33, 58
src/axelrod_dojo/archetypes/__init__.py                          0      0   100%
src/axelrod_dojo/archetypes/cycler.py                           49      0   100%
src/axelrod_dojo/archetypes/fsm.py                             111      2    98%   87, 89
src/axelrod_dojo/archetypes/gambler.py                          33      0   100%
src/axelrod_dojo/archetypes/hmm.py                             150      0   100%
src/axelrod_dojo/utils.py                                      123     14    89%   38, 52, 188-200
src/axelrod_dojo/version.py                                      1      0   100%
------------------------------------------------------------------------------------------
TOTAL                                                          599     20    97%
```

drvinceknight commented:
> The build is failing from tests in other areas than the Cycler:

Not quite :) What you've pasted there is the coverage results (which lines of the source code are covered by tests). There are a number of integration tests that are failing indicating something else that's broken.

GitToby committed:
removed GA output header

GitToby committed:
cleaned imports for GA

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+
+        Parameters
+        ----------
+        sequence_length - length or random moves to generate
```
or -> of

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+    @staticmethod
+    def generate_random_sequence(sequence_length):
+        """
+        Generates a sequence of random moves when an instance is initialised
```
remove "when an instance is initialized"

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+    def __init__(self, sequence=None, sequence_length: int = 200, mutation_probability=0.1, mutation_potency=1):
+
+        if sequence is None:
+            # generates sequence uses a map to the Action class to change a sequence of 1 & 0s to C & D actions
```
Sentence is confusing

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+
+        Returns
+        -------
+        list - a list of C & D actions: e.g. [C,C,D,D,C]
```
`list[Action]`

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+        -------
+        list - a list of C & D actions: e.g. [C,C,D,D,C]
+        """
+        return list(map(axl.Action, random.randint(0, 1 + 1, (sequence_length, 1))))
```
2?

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+
+        midpoint = int(len(seq1) / 2)
+
+        # If the length is odd, keep more from sequence 1
```
Does this introduce bias of some kind? Might be better to randomly choose which sequence to keep more of.

GitToby replied:
any tips on an elegant way of testing a randomly generated split? Currently Ive just seeded it in the tests

drvinceknight replied:
Seeding the tests is the way to go :+1: (apologies if this comment is late or not helpful, connectivity is patchy at the moment so dipping in and out of things).

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+            for _ in range(self.mutation_potency):
+                index_to_change = random.randint(0, len(mutated_sequence))
+                # Mutation - change a single gene
+                if mutated_sequence[index_to_change] == C:
```
I think we have a flip_dict from Axelrod that maps C -> D and D -> C

drvinceknight replied:
They're methods on the instances:

```python
>>> import axelrod as axl
>>> C = axl.Action.C
>>> C.flip()
D
```

marcharper replied:
Even better :)

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+        -------
+        CyclerParams - a separate instance copy of itself.
+        """
+        # seq length will be provided when copying
```
Remove this comment

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+
+        return string_sequence
+
+    # Getters --------------------
```
In Python we generally don't write getter functions for classes unless they are needed for property decorators. It's fine to just directly reference the instance variables elsewhere.

marcharper commented on src/axelrod_dojo/utils.py:
```diff
@@ -11,17 +10,20 @@
-        self.output.flush()
-        os.fsync(self.output.fileno())
+    def write_row(self, row):
+        with open(self.file, self.mode, newline='') as file_writer:
```
Are you sure that you want no newline here?

GitToby replied:
it makes the csv cleaner on output, I can change it back if needed. The solve was from [this](https://stackoverflow.com/questions/3191528/csv-in-python-adding-an-extra-carriage-return) thread on having too many carriage returns on windows.

marcharper commented on src/axelrod_dojo/utils.py:
```diff
@@ -11,17 +10,20 @@
+            writer.writerow(row)
+
+    def clear_file(self):
+        with open(self.file, 'w+', newline='') as file_writer:
```
and here?

GitToby replied:
As above, this would add a new line on the "empty" file.

marcharper commented on src/axelrod_dojo/utils.py:
```diff
@@ -11,17 +10,20 @@
+            writer = csv.writer(file_writer)
+            writer.writerow(row)
+
+    def clear_file(self):
```
Just pass in `mode='w'` into `__init__`

GitToby replied:
I dont understand this comment, is this for the `clear_file()` method?

marcharper replied:
Yes, it's not necessary to have a separate function to clear the file. If you open with flag 'w' instead of 'a' the file will be cleared.

marcharper replied:
Is the `clear_file` function still being used?

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -1,26 +1,32 @@
         self.params_class = params_class
         self.bottleneck = bottleneck
-
+        self.print_output = print_output
```
Consider making `print_output` an argument to `evolve` instead of a instance variable.

GitToby replied:
I was thinking that eventually it could propagate through as a sort of verbosity level (print each cycle (or every n cycles) or each game etc.). Moving this to a method param would also be harder to control over multiple generation cycles; in my opinion using a getter & setter would be the easiest way to control the output levels.

marcharper replied:
Ok, I don't feel strongly about this. Regardless there's no need for a getter or setter.

drvinceknight replied:
> I was thinking that eventually it could propagate through as a sort of verbosity level (print each cycle (or every n cycles) or each game etc.)

In general I find "coding for the future" introduces complexities that don't often end up being used but need to be maintained. I'd prefer this to follow @marcharper 's suggestion :+1:

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+
+        return CyclerParams(sequence=new_seq)
+
+    def mutate(self):
```
Is it possible for the sequence to change length?

GitToby replied:
its not possible to change the length from within the crossover method. bug #43 was solved in the the `Population` class and every generation should have the same length now

GitToby committed:
made changes to comments & crossover & mutation as requested in pr

marcharper commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -13,15 +13,15 @@ class CyclerParams(Params):
 
     **********
-    TODOs:
+    Expansions:
```
If you don't plan to do this in the current CL please remove the comment. You can open an issue in the issue tracker instead.

drvinceknight commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+        return string_sequence
+
+    # Getters --------------------
+    def get_sequence(self):
```
We don't really need all these methods.

drvinceknight replied:
Woops, I see that this is what @marcharper's comment was (sorry viewing this in the overall conversation view and the code review tab confused me). :+1:

drvinceknight commented on tests/archetypes/test_cycler.py:
```diff
@@ -0,0 +1,71 @@
+    # Basic creation methods setting the correct params
+
+    def test_creation_seqLen(self):
+        # TODO: create a particular creation separate to length
```
Clear up this `# TODO`

drvinceknight commented on src/axelrod_dojo/archetypes/cycler.py:
```diff
@@ -0,0 +1,142 @@
+        """
+        Basic mutation which may change any random gene(s) in the sequence.
+        """
+        # if the mutation occurs
```
No need for this inline comment.

drvinceknight reviewed (changes requested):
Minor comments from me that we can discuss in our meeting today @GitToby (essentially just reaffirming what @marcharper has said).

GitToby committed:
Made changes requested in the PR

Nikoleta-v3 commented:
Hey @GitToby it looks like you are still using the `clear_file` function in your `TestOutputer`.

```
======================================================================
ERROR: test_write_and_clear (unit.test_utils.TestOutputer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/Axelrod-Python/axelrod-dojo/tests/unit/test_utils.py", line 26, in test_write_and_clear
    self.outputer.clear_file()
AttributeError: 'Outputer' object has no attribute 'clear_file'
---------------------------------------------------------------------
```

GitToby committed:
Removed Test for removed functionality

GitToby commented:
CI is failing due to tests on the form of the output; I'll change it back after analysis round has finished

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+
+        # Opponent Sequence can be reconstructed by playing a match against the exact seed params
+
+        row = [self.generation, # 0
```
Comments with numbers here are not necessary since it's defined above

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+        # Row Data Cols:
+        # 0:Generation no.
+        # 1:Mean score over pop
+        # 2:Medan score over pop
```
Median

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+                pass
+
+        # Row Data Cols:
+        # 0:Generation no.
```
number

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+
+        # Row Data Cols:
+        # 0:Generation no.
+        # 1:Mean score over pop
```
population

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+               mean(scores),# 1
+               median(scores),# 2
+               pstdev(scores),# 3
+               results[0][0]-results[len(results)-1][0],# 4
```
spaces around `-`

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+        # 3:Population Variance
+        # 4:Score range
+        # 5:Highest scoring result
+        # 6:sequence of highest scoring result
```
Sequence --> parameters (it's not necessarily a sequence)

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+        for strategy, init_kwargs in self.opponents_information:
+            opponent = strategy(**init_kwargs)
+            opponent_type = type(opponent).name
+            player_seed = None
```
Move this line into the exception clause and remove the `pass`

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+        # Report
+        if self.print_output:
+            print()
+            print(row)  # prints best result
```
This comment isn't accurate (or necessary)

marcharper commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -92,18 +92,46 @@ def evolve(self):
+
+        for strategy, init_kwargs in self.opponents_information:
+            opponent = strategy(**init_kwargs)
+            opponent_type = type(opponent).name
```
opponent_strategy_name

GitToby commented:
@marcharper @drvinceknight  The changes I made are causing the tests to fail; what direction do you want me to go in? I feel like the unpacking of results in tests is causing an issue with customizing the code to what is needed in individual cases. For now i'll move my customized code into a new class but I suggest that there be some sort of contribution guide to forking the repo to include a note like 'remember to extend any classes you want to personalise' with a short guide to help new PRs get through. if you think this is a good idea, where should I add it? if not should I just revert my code after im done?

marcharper commented:
More documentation is always great to have :) It's normal to run into these kinds of issues in the first contribution. You're almost there!

The tests are failing because of:
```
Traceback (most recent call last):
  File "/home/travis/build/Axelrod-Python/axelrod-dojo/tests/integration/test_fsm.py", line 49, in test_score
    _, mean_score, sd_score, max_score, arg_max = row
ValueError: too many values to unpack (expected 5)
```
This is probably because of the addition of median to row. If you fix the unpacking here the tests should be ok.

Also: it might be a good idea to turn off per round reporting for the tests. Nice to see the output but makes the test results harder to look through.

GitToby commented:
Sure, Ill make the changes and alter the unpacking in failed tests. where should i make changes to the docs?

GitToby commented:
@drvinceknight  @marcharper How can i fix the doctests? I dont understand whats going wrong

drvinceknight commented:
I personally agree and don't think the changes you've made to unpack more output should go in to the library. I think that's something very specific to your use case for your project. What do you think @marcharper?

We can/should write some info about contributing but let's just open that as another issue and do it at a later date to keep this PR clean.

drvinceknight commented:
> @drvinceknight @marcharper How can i fix the doctests? I dont understand whats going wrong

This is the stacktrace:

```
File "./docs/tutorial/index.rst", line 44, in index.rst
Failed example:
    population.run(generations)
Expected:
    Scoring Generation 1
    Generation 1 | Best Score: 2.1 0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D
    Scoring Generation 2
    Generation 2 | Best Score: 2.1 0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D
    Scoring Generation 3
    Generation 3 | Best Score: 2.1 0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D
    Scoring Generation 4
    Generation 4 | Best Score: 2.1 0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D
Got:
    Starting evolution for 4 generations
    Scoring Generation 1
    [1, 1.77, 1.7666666666666668, 0.19969421067666884, 0.6000000000000001, 2.1, '0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D', 'Defector', None]
    <BLANKLINE>
    Scoring Generation 2
    [2, 1.8166666666666667, 1.7333333333333334, 0.23225943348859798, 0.6000000000000001, 2.1, '0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D', 'Defector', None]
    <BLANKLINE>
    Scoring Generation 3
    [3, 1.9366666666666668, 2.0833333333333335, 0.21804688588568394, 0.6000000000000001, 2.1, '0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D', 'Defector', None]
    <BLANKLINE>
    Scoring Generation 4
    [4, 1.96, 2.1, 0.17876116903722566, 0.46666666666666656, 2.1, '0:C:0_C_0_C:0_D_1_C:1_C_1_D:1_D_1_D', 'Defector', None]
    <BLANKLINE>
    Done.
```

So what is currently written in the `"./docs/tutorial/index.rst"` file is not what your code outputs (because you've added more to stdout).

drvinceknight commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -77,22 +84,57 @@ def crossover(population, num_variants):
-        self.outputer.write(row)
-
-        ## Next Population
+        # Note: if using this for analysis, for reproducability it may be useful to alter the data passed back
```
Let us remove this inline comment.

drvinceknight commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -77,22 +84,57 @@ def crossover(population, num_variants):
+            except KeyError:
+                player_seed = None
+
+        # Row Data Cols:
```
This risk becoming lies when/if something gets added, I think what is in each column is pretty explicit so let's just remove the inline comments.

drvinceknight replied:
A better approach would be to create a temporary variable for some of the columns (the ones that need it perhaps):

```
population_score_range = results[0][0] - results[len(results) - 1][0]
```

and then put that in `row`.

drvinceknight reviewed (changes requested):
Two minor requests to remove some inline comments assuming the output stays as it.

drvinceknight commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -77,22 +84,57 @@ def crossover(population, num_variants):
+        # Report to log
+        if self.print_output:
+            print(row)
+            print()
```
Let's remove this blank line.

drvinceknight commented on src/axelrod_dojo/algorithms/genetic_algorithm.py:
```diff
@@ -77,22 +84,57 @@ def crossover(population, num_variants):
+        # 8:Seed (if applicable, else None)
+
+        # Opponent Sequence can be reconstructed by playing a match against the exact seed params
+        row = [self.generation,
```
So this is the reason the doctest is failing, you're outputting way more information then is currently in the doctest which is expecting a different output. If we want to keep this extra output then the doctest should maybe be run with `print_output=False` because the floats will continue to cause errors even if you wrote them in.

This would require changing the rest of that tutorial too though obviously (perhaps just explaining that the output has been written to a `training_output.csv`).

drvinceknight commented:
My suggestion would be that @GitToby creates another branch for his work with all current changes and for the purpose of this PR reverts this branch to 7a95ca6 (which has already passed CI).

Thus this PR would be an implementation of Cycler as well as the fix of #43.

(**EDIT** We can open issues to keep track of adding docs and adding to the output :+1:)

GitToby commented:
@marcharper, @drvinceknight and I had a chat about the scope creep in this PR and we feel its best to close it as soon as we can. I will make changes to the test logging and the print parameters in the `Population` class but after that I don't think any more changes are needed; if I missed anything let me know but minor features should probably be opened as an issue and we can pick them up on another branch.

GitToby committed:
updated test logging

 - Moved print_output to run method
 - updated docs
 - removed logging in tests to make CI readable

drvinceknight commented:
I'm happy with this and have no further requests. Nice job @GitToby.

marcharper commented:
Thanks for the contribution!
